File: src/audit/signing/kmsRetry.ts
// Copyright (c) 2026 dotandev
// SPDX-License-Identifier: MIT OR Apache-2.0

/**
 * Retry configuration, error classification, and idempotency cache for
 * the AWS KMS Signer (Issue 66).
 *
 * Design rationale lives alongside the Go implementation in
 * internal/signer/kms_retry.go; the two are kept structurally similar
 * so future audits can compare them line-for-line.
 */

// ---------------------------------------------------------------------------
// Retry configuration
// ---------------------------------------------------------------------------

export interface KmsRetryConfig {
  /** Max attempts beyond the initial call. 0 disables retries. Default 3. */
  maxRetries: number;
  /** Initial backoff in milliseconds before the first retry. Default 250. */
  initialBackoffMs: number;
  /** Upper bound for a single backoff interval in ms. Default 5000. */
  maxBackoffMs: number;
  /** Proportional jitter added to each backoff (0 disables). Default 0.2. */
  jitterFraction: number;
  /** Client-side signature cache TTL in ms. 0 disables caching. Default 60000. */
  idempotencyTtlMs: number;
  /** Maximum entries in the LRU cache. Default 1024. */
  idempotencyMaxEntries: number;
}

/**
 * Conservative production defaults for the KMS Signer: 3 retries,
 * exponential backoff 250ms → 5s with ±20% jitter, and a 60-second
 * client-side idempotency window sized for ~1k unique digests. These
 * values mirror internal/signer.DefaultKMSRetryConfig exactly so the
 * Go CLI and the Node SDK apply the same knobs.
 */
export function defaultKmsRetryConfig(): KmsRetryConfig {
  return {
    maxRetries: 3,
    initialBackoffMs: 250,
    maxBackoffMs: 5000,
    jitterFraction: 0.2,
    idempotencyTtlMs: 60_000,
    idempotencyMaxEntries: 1024,
  };
}

/**
 * Reads GLASSBOX_KMS_* overrides from the current process environment.
 * Unknown / malformed values are silently ignored so a typo in the
 * environment cannot cascade into a runtime error during signing.
 */
export function loadKmsRetryConfigFromEnv(
  base: KmsRetryConfig = defaultKmsRetryConfig(),
): KmsRetryConfig {
  const env = process.env;
  const cfg: KmsRetryConfig = { ...base };

  if (env.GLASSBOX_KMS_MAX_RETRIES != null) {
    const n = Number(env.GLASSBOX_KMS_MAX_RETRIES);
    if (Number.isFinite(n) && n >= 0) cfg.maxRetries = n;
  }
  if (env.GLASSBOX_KMS_INITIAL_BACKOFF_MS != null) {
    const n = Number(env.GLASSBOX_KMS_INITIAL_BACKOFF_MS);
    if (Number.isFinite(n) && n >= 0) cfg.initialBackoffMs = n;
  }
  if (env.GLASSBOX_KMS_MAX_BACKOFF_MS != null) {
    const n = Number(env.GLASSBOX_KMS_MAX_BACKOFF_MS);
    if (Number.isFinite(n) && n >= 0) cfg.maxBackoffMs = n;
  }
  if (env.GLASSBOX_KMS_JITTER_FRACTION != null) {
    const n = Number(env.GLASSBOX_KMS_JITTER_FRACTION);
    if (Number.isFinite(n) && n >= 0) cfg.jitterFraction = n;
  }
  if (env.GLASSBOX_KMS_IDEMPOTENCY_TTL_MS != null) {
    const n = Number(env.GLASSBOX_KMS_IDEMPOTENCY_TTL_MS);
    if (Number.isFinite(n) && n >= 0) cfg.idempotencyTtlMs = n;
  }
  if (env.GLASSBOX_KMS_IDEMPOTENCY_MAX != null) {
    const n = Number(env.GLASSBOX_KMS_IDEMPOTENCY_MAX);
    if (Number.isFinite(n) && n > 0) cfg.idempotencyMaxEntries = n;
  }

  return cfg;
}

// ---------------------------------------------------------------------------
// SignMetadata
// ---------------------------------------------------------------------------

export interface SignMetadata {
  /** Signature bytes on success. Undefined on error. */
  signature?: Buffer;
  /** Total number of KMS API calls; 0 when the idempotency cache hit. */
  attempts: number;
  /** Opaque caller-supplied id threaded through every attempt. */
  correlationId?: string;
  /** True when the signature was returned from the in-memory cache. */
  idempotencyHit: boolean;
  /** AWS error code (or "NetworkError", "ContextCancelled", "EmptyMessage"). Undefined on success. */
  errorCode?: string;
  /** Coarse category: "api" | "network" | "context" | "input" | "unknown". Undefined on success. */
  errorClass?: string;
  /** True if the final error was classified as retryable. */
  retryable: boolean;
  /** Wall-clock duration of the entire call in ms. */
  elapsedMs: number;
}

export class KmsEmptyMessageError extends Error {
  constructor() {
    super("kms signer: empty message");
    this.name = "KmsEmptyMessageError";
  }
}

export class KmsContextCancelledError extends Error {
  constructor(public readonly cause: unknown) {
    super("kms signer: context cancelled");
    this.name = "KmsContextCancelledError";
  }
}

// ---------------------------------------------------------------------------
// Error classification
// ---------------------------------------------------------------------------

/**
 * AWS KMS error codes classified as always safe to retry. Mirrors the
 * Go retryableKMSErrorCodes map in internal/signer/kms_retry.go.
 */
const RETRYABLE_KMS_ERROR_CODES: ReadonlySet<string> = new Set<string>([
  "InternalError",
  "ServiceUnavailable",
  "ThrottlingException",
  "TooManyRequests",
  "TooManyRequestsException",
  "RequestTimeout",
  "RequestTimeoutException",
  "KMSInternalException",
  "UnavailableException",
  "RequestLimitExceeded",
  "ProvisionedThroughputExceededException",
]);

/**
 * Exception names thrown by the AWS SDK v3 that count as retryable
 * transport failures (DNS, connection reset, socket hangup, etc.). The
 * SDK v3 uses a unified `error.name` convention.
 */
const RETRYABLE_TRANSPORT_NAMES: ReadonlySet<string> = new Set<string>([
  "TimeoutError",
  "ECONNRESET",
  "ECONNREFUSED",
  "ENOTFOUND",
  "EAI_AGAIN",
  "EPIPE",
  "ETIMEDOUT",
  "NetworkingError",
]);

function isRetryableKmsErrorCode(code: string): boolean {
  return RETRYABLE_KMS_ERROR_CODES.has(code);
}

/**
 * Inspects an SDK error and returns (retryable, code, class).
 * Never records nor logs payload bytes — only the SDK-reported code or
 * the originating Error's name.
 */
export function classifyKmsError(err: unknown): {
  retryable: boolean;
  code: string;
  class: "api" | "network" | "unknown";
} {
  if (err == null) return { retryable: false, code: "", class: "unknown" };

  if (err instanceof Error) {
    const name = err.name ?? "";
    if (RETRYABLE_TRANSPORT_NAMES.has(name)) {
      return { retryable: true, code: "NetworkError", class: "network" };
    }
    const anyErr = err as Error & { code?: string; $metadata?: unknown };
    // Prefer an explicit AWS SDK error code; only fall back to the
    // Error's `name` when it carries domain meaning (i.e. it is not
    // the generic "Error" name a plain `new Error(...)` produces).
    // This keeps opaque JS errors from being misclassified as API
    // errors just because Error.prototype.name === "Error".
    const explicitCode = anyErr.code;
    const code = explicitCode ?? (name && name !== "Error" ? name : "");
    if (code) {
      return {
        retryable: isRetryableKmsErrorCode(code),
        code,
        class: "api",
      };
    }
  }
  return { retryable: false, code: "Unknown", class: "unknown" };
}

// ---------------------------------------------------------------------------
// Backoff
// ---------------------------------------------------------------------------

/**
 * Returns the next backoff in ms using exponential growth capped at
 * maxBackoffMs, plus optional proportional jitter. `jitterUnit` is a
 * function that returns a uniform random number in [-1, 1]; tests
 * inject a deterministic source to keep assertions reproducible.
 */
export function nextRetryBackoffMs(
  cfg: KmsRetryConfig,
  currentMs: number,
  jitterUnit: () => number,
): number {
  const base = currentMs <= 0 ? cfg.initialBackoffMs : currentMs * 2;
  const capped = base > cfg.maxBackoffMs ? cfg.maxBackoffMs : base;
  if (cfg.jitterFraction <= 0 || jitterUnit == null) {
    return capped;
  }
  const j = jitterUnit();
  const delta = capped * cfg.jitterFraction * j;
  const adjusted = capped + delta;
  return adjusted < 0 ? 0 : adjusted;
}

/**
 * Sleeps for ms, or aborts when an AbortSignal fires. Returns true on
 * full sleep, false on abort (with reason available via the signal).
 */
export function sleepWithAbort(ms: number, signal?: AbortSignal): Promise<boolean> {
  if (ms <= 0) return Promise.resolve(signal?.aborted !== true);
  return new Promise((resolve) => {
    let done = false;
    const finish = () => {
      if (done) return;
      done = true;
      if (timer) clearTimeout(timer);
      resolve(!signal?.aborted);
    };
    const timer = setTimeout(finish, ms);
    if (signal) {
      if (signal.aborted) {
        finish();
        return;
      }
      signal.addEventListener("abort", finish, { once: true });
    }
  });
}

// ---------------------------------------------------------------------------
// safeKeyIdRef — log-safe identifier
// ---------------------------------------------------------------------------

/**
 * Returns a short suffix of a KMS key id suitable for log lines. Empty
 * values become "(none)" so log lines are unambiguous. Full key ids
 * (ARNs, aliases) are not AWS secrets but are still identifiers; we
 * truncate to a fixed-length suffix so log scraping is less valuable
 * to an attacker with logs but no production credentials.
 */
export function safeKeyIdRef(keyId: string | undefined | null): string {
  if (!keyId) return "(none)";
  const suffixLen = 8;
  if (keyId.length <= suffixLen) return keyId;
  return "..." + keyId.slice(-suffixLen);
}

// ---------------------------------------------------------------------------
// Idempotency cache (TTL + LRU)
// ---------------------------------------------------------------------------

/**
 * Tiny TTL+LRU table keyed by opaque strings. Used to short-circuit
 * identical re-signs within the configured TTL window without paying
 * another KMS bill and another client-side Encrypt call. All methods
 * are safe for concurrent use.
 *
 * Privacy: the cache stores signature bytes only — never messages,
 * digests, or key ids. The lookup key binds key id and a SHA-256 of
 * the message but never the plain message.
 */
export class IdempotencyCache {
  private readonly entries = new Map<string, { value: Uint8Array; expiresAt: number }>();
  private readonly maxEntries: number;
  private readonly ttlMs: number;
  private readonly now: () => number;

  constructor(maxEntries: number, ttlMs: number, now: () => number = Date.now) {
    this.maxEntries = maxEntries;
    this.ttlMs = ttlMs;
    this.now = now;
  }

  /** Returns true when caching is effectively disabled. */
  get disabled(): boolean {
    return this.ttlMs <= 0 || this.maxEntries <= 0;
  }

  get(key: string): Uint8Array | undefined {
    if (this.disabled) return undefined;
    const entry = this.entries.get(key);
    if (!entry) return undefined;
    if (this.now() > entry.expiresAt) {
      this.entries.delete(key);
      return undefined;
    }
    // LRU touch — re-insert to move to back of insertion order.
    this.entries.delete(key);
    this.entries.set(key, entry);
    return entry.value;
  }

  put(key: string, value: Uint8Array): void {
    if (this.disabled || !key) return;
    // Clone to insulate from caller-side mutation.
    const clone = new Uint8Array(value.byteLength);
    clone.set(value);
    if (this.entries.has(key)) this.entries.delete(key);
    else if (this.entries.size >= this.maxEntries) {
      // Evict oldest (first inserted).
      const oldestKey = this.entries.keys().next().value;
      if (oldestKey !== undefined) this.entries.delete(oldestKey);
    }
    this.entries.set(key, { value: clone, expiresAt: this.now() + this.ttlMs });
  }

  size(): number {
    return this.entries.size;
  }
}

/**
 * Cached reference to Node's `crypto` module so we don't pay the
 * eval-require cost on every idempotency-key derivation. The cache
 * is process-global: only one signer per process should ever import
 * Node crypto, so this is safe even across multiple KmsSigner instances.
 */
let cachedNodeCrypto: any | undefined;

function loadNodeCrypto(): any {
  if (!cachedNodeCrypto) {
    // eslint-disable-next-line @typescript-eslint/no-var-requires, no-eval
    cachedNodeCrypto = eval("require")("crypto");
  }
  return cachedNodeCrypto;
}

/**
 * Computes the idempotency cache key for a (key id, message) pair.
 * The message is hashed so the raw payload never appears in the cache
 * key string; excludes chance of the key surfacing in process listings
 * or breaker traces. The `v1:` prefix lets us rev the key scheme later
 * without colliding with cached entries.
 */
export function idempotencyKey(
  keyId: string,
  message: Uint8Array,
): string {
  const nodeCrypto = loadNodeCrypto();
  const hash = nodeCrypto.createHash("sha256");
  hash.update(Buffer.from(message));
  return `v1:${keyId}:${hash.digest("hex")}`;
}

/**
 * Reset the cached Node crypto reference. Test-only escape hatch so
 * tests can install a crypto polyfill before the first derivation.
 */
export function _resetIdempotencyCryptoForTesting(): void {
  cachedNodeCrypto = undefined;
}

// ---------------------------------------------------------------------------
// Logging hook
// ---------------------------------------------------------------------------

export interface KmsLogFunction {
  (level: "debug" | "warn", msg: string, attrs: Record<string, unknown>): void;
}

/**
 * Default log function: writes debug-and-above events through the
 * global console as a single readable line; production deployments
 * are expected to swap this for an OpenTelemetry / structured-log
 * adapter via the constructor option.
 */
export function defaultKmsLog(logger: KmsLogFunction = consoleLog): KmsLogFunction {
  return logger;
}

function consoleLog(
  level: "debug" | "warn",
  msg: string,
  attrs: Record<string, unknown>,
): void {
  // Never include payload bytes, full key ids, or signature bytes in
  // the formatted text — only known-safe, scalar attributes.
  const entries = Object.entries(attrs)
    .map(([k, v]) => `${k}=${safeFormat(v)}`)
    .join(" ");
  const line = `[kms ${level}] ${msg}${entries ? " " + entries : ""}`;
  if (level === "warn") {
    console.warn(line);
  } else {
    console.debug(line);
  }
}

function safeFormat(v: unknown): string {
  if (v == null) return "null";
  if (typeof v === "number" || typeof v === "boolean") return String(v);
  if (typeof v === "string") {
    // Bound log-line size & avoid leaking large strings.
    return v.length > 64 ? v.slice(0, 64) + "..." : v;
  }
  return "[complex]";
}

File: internal/signer/kms_retry.go
// Copyright 2026 Glassbox Users
// SPDX-License-Identifier: Apache-2.0

package signer

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"net"
	"time"

	"github.com/aws/smithy-go"
)

// KMSRetryConfig configures bounded retries and client-side idempotency
// for the AWS KMS-backed Signer.
//
// All fields are optional and have safe defaults; configure explicitly
// when running in environments known to be flaky or when stricter
// guarantees are required (e.g. shorter caps to bound worst-case latency).
type KMSRetryConfig struct {
	// MaxRetries bounds the total number of retry attempts on a
	// retryable error. 0 disables retries (single-shot, current
	// pre-Issue-66 behaviour).
	MaxRetries int

	// InitialBackoff is the wait before the first retry. Subsequent
	// waits double (exponential), capped at MaxBackoff, with optional
	// proportional jitter.
	InitialBackoff time.Duration

	// MaxBackoff is the upper bound for any single backoff interval.
	MaxBackoff time.Duration

	// JitterFraction is a multiplier (0 = none, 0.1 = ±10%) added to
	// each calculated backoff so concurrent retriers do not synchronise.
	JitterFraction float64

	// IdempotencyTTL caches a successful signature keyed by
	// (keyId, canonical digest) for this duration. 0 disables caching.
	IdempotencyTTL time.Duration

	// IdempotencyMaxEntries caps the in-memory cache size; entries are
	// evicted LRU once the cap is exceeded. 0 effectively disables
	// caching regardless of IdempotencyTTL.
	IdempotencyMaxEntries int
}

// DefaultKMSRetryConfig returns a conservative, production-friendly
// retry policy: at most 3 retry attempts with exponential backoff
// doubled from 250ms up to 5s and 20% jitter, plus a 60-second
// client-side idempotency window with a 1k-entry LRU cap.
//
// Tunables follow Issue 66 acceptance criteria: bounded retries on
// safe failures, identical digest across attempts, no payload logging
// (configurable only via env) and at most one KMS call per audit log
// even if callers race.
func DefaultKMSRetryConfig() KMSRetryConfig {
	return KMSRetryConfig{
		MaxRetries:            3,
		InitialBackoff:        250 * time.Millisecond,
		MaxBackoff:            5 * time.Second,
		JitterFraction:        0.2,
		IdempotencyTTL:        60 * time.Second,
		IdempotencyMaxEntries: 1024,
	}
}

// KMSSignMetadata captures the observable outcome of a single KMS Sign
// call. The classic Sign(data) returns only the bytes; SignWithMetadata
// exposes retry/idempotency details so callers (CLI, tests, telemetry)
// can introspect the path.
type KMSSignMetadata struct {
	// Signature is the produced bytes. Nil on error.
	Signature []byte

	// Attempts is the total number of KMS API calls (1 on success with
	// no retries, more after transient retries, 0 on idempotency hit).
	Attempts int

	// CorrelationID is the opaque id threaded through every attempt for
	// traceability. May be empty.
	CorrelationID string

	// IdempotencyHit is true when the signature was returned from the
	// in-memory cache without contacting KMS.
	IdempotencyHit bool

	// ErrorCode is the AWS API error code (or "NetworkError",
	// "ContextCancelled", "EmptyMessage") on failure. Empty on success.
	ErrorCode string

	// ErrorClass is a coarse category: "api", "network", "context",
	// "input", "unknown". Empty on success.
	ErrorClass string

	// Retryable reports whether the final error was classified as
	// retryable. Always false on success and on the very last attempt.
	Retryable bool

	// Elapsed is the wall-clock duration of the whole call (cache lookup
	// + retry loop). Useful for latency budgets and tests.
	Elapsed time.Duration
}

// ErrEmptyMessage is returned by SignWithMetadata when the caller passes
// an empty input. AWS KMS rejects empty messages with a confusing
// "must not be empty" error; surfacing a typed sentinel lets callers
// pre-flight cheaply and tests assert on identity.
var ErrEmptyMessage = errors.New("kms signer: empty message")

// ErrContextCancelled is returned when the supplied context is done
// before all retry attempts complete. The underlying ctx.Err() is wrapped
// so callers still see the precise cancel cause.
var ErrContextCancelled = errors.New("kms signer: context cancelled")

// retryableKMSErrorCodes is the allowlist of AWS KMS error codes that
// are always safe to retry. Mirrors the SDK's retryable classification
// but is owned here so it can be asserted against in tests and matched
// deterministically regardless of SDK defaults.
var retryableKMSErrorCodes = map[string]struct{}{
	"InternalError":                          {},
	"ServiceUnavailable":                     {},
	"ThrottlingException":                    {},
	"TooManyRequests":                        {},
	"TooManyRequestsException":               {},
	"RequestTimeout":                         {},
	"RequestTimeoutException":                {},
	"KMSInternalException":                   {},
	"UnavailableException":                   {},
	"RequestLimitExceeded":                   {},
	"ProvisionedThroughputExceededException": {},
}

// isRetryableKMSCode returns true when the AWS SDK-reported code is on
// the safe-to-retry allowlist.
func isRetryableKMSCode(code string) bool {
	if code == "" {
		return false
	}
	_, ok := retryableKMSErrorCodes[code]
	return ok
}

// classifyKMSError inspects an SDK error and reports whether it can be
// retried, the SDK error code (or "NetworkError"), and a coarse class.
//
// The function never returns, logs, or buffers payload bytes — it only
// extracts the SDK-reported error code via smithy.APIError.
func classifyKMSError(err error) (retryable bool, code string, class string) {
	if err == nil {
		return false, "", ""
	}

	// Network-layer failures (timeouts, connection refused) are
	// retryable. Smithy wraps these into smithy.OperationError; we
	// unwrap to find a stdlib net.Error.
	var nerr net.Error
	if errors.As(err, &nerr) {
		return true, "NetworkError", "network"
	}

	var apiErr smithy.APIError
	if errors.As(err, &apiErr) {
		code = apiErr.ErrorCode()
		return isRetryableKMSCode(code), code, "api"
	}

	// Unknown / opaque error: do not retry (safer default). Tests
	// assert this behaviour so future changes cannot silently widen
	// the retry surface.
	return false, "Unknown", "unknown"
}

// safeKeyIDRef returns a short, non-secret suffix of a KMS key id
// suitable for log lines. Full key ids are not secrets in AWS but are
// still identifiers; truncating makes log scraping less useful to an
// attacker who has access to logs but not the production credential
// store. Empty key ids are mapped to "(none)" so log lines are unambiguous.
func safeKeyIDRef(keyID string) string {
	if keyID == "" {
		return "(none)"
	}
	const suffixLen = 8
	if len(keyID) <= suffixLen {
		return keyID
	}
	return "..." + keyID[len(keyID)-suffixLen:]
}

// nextRetryBackoff returns the next wait duration using exponential
// backoff with optional proportional jitter, capped at cfg.MaxBackoff.
//
// When current is zero (the first call) the function returns
// cfg.InitialBackoff so the seeded value seeds the doubling curve
// instead of doubling on top of itself. Subsequent calls double the
// previous backoff, capped at cfg.MaxBackoff. Jitter = uniform *
// JitterFraction * next. Tests inject a deterministic jitter source to
// keep assertions stable.
func nextRetryBackoff(cfg KMSRetryConfig, current time.Duration, jitterSrc func() float64) time.Duration {
	if current <= 0 {
		return cfg.InitialBackoff
	}
	next := current * 2
	if next > cfg.MaxBackoff {
		next = cfg.MaxBackoff
	}
	if cfg.JitterFraction > 0 && jitterSrc != nil {
		j := jitterSrc()
		// Clamp jitter to [-1, 1] so callers can pass sources that
		// produce out-of-range values (e.g. [-1, 1] vs [0, 1]).
		if j < -1 {
			j = -1
		}
		if j > 1 {
			j = 1
		}
		delta := float64(next) * cfg.JitterFraction * j
		adj := float64(next) + delta
		if adj < 0 {
			adj = 0
		}
		return time.Duration(adj)
	}
	return next
}

// waitWithContext sleeps for d or returns ctx.Err() if the context is
// cancelled first.
func waitWithContext(ctx context.Context, d time.Duration) error {
	if d <= 0 {
		return ctx.Err()
	}
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-t.C:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

// idempotencyKey builds the cache key used by the in-memory idempotency
// cache. It mixes the key id (so changing key ids does not return a
// stale signature for the wrong key) and a SHA-256 of the message
// (so the full digest is never part of the key string, including in
// test logging). The "v1:" prefix lets us rev the key scheme later.
func idempotencyKey(keyID string, message []byte) string {
	sum := sha256.Sum256(message)
	return "v1:" + keyID + ":" + hex.EncodeToString(sum[:])
}

// logKMSSign is the structured-log hook used by KMSSigner. Tests inject
// a stub; production code uses logger.Logger.Debug by default. The
// signature intentionally takes a variadic any so we cannot
// accidentally accept the full message/digest — production callers only
// pass named attributes.
type logKMSSign func(level logLevel, msg string, attrs ...any)

type logLevel int

const (
	logDebug logLevel = iota
	logWarn
)

// defaultKMSLog returns the production log hook: debug-level events
// are routed to the package logger; warn-level events are routed too.
// The function intentionally does NOT take ctx — KMSRetryConfig is
// loggable without a parent context, and correlation ID flows via
// metadata attribute (already a string) rather than context value.
func defaultKMSLog() logKMSSign {
	// Lazy import to avoid a hard dependency on the logger package
	// for tests that wire a stub. internal/logger.Logger is initialised
	// at process start and is safe to use.
	return func(level logLevel, msg string, attrs ...any) {
		// Avoid pulling internal/logger into every test binary by
		// indirection. logger.Logger is *slog.Logger; we adapt here.
		logKMSSignToLogger(level, msg, attrs...)
	}
}
